go/vt/vterrors/vterrors.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2012, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package vterrors provides helpers for propagating internal errors through the Vitess
+// system (including across RPC boundaries) in a structured way.
+package vterrors
+
+import (
+	"fmt"
+
+	mproto "github.com/youtube/vitess/go/mysql/proto"
+)
+
+const (
+	// TabletError is the base VtTablet error. All VtTablet errors should be 4 digits, starting with 1.
+	TabletError = 1000
+	// UnknownTabletError is the code for an unknown error that came from VtTablet.
+	UnknownTabletError = 1999
+	// VtgateError is the base VTGate error code. All VTGate errors should be 4 digits, starting with 2.
+	VtgateError = 2000
+	// UnknownVtgateError is the code for an unknown error that came from VTGate.
+	UnknownVtgateError = 2999
+)
+
+// VitessError is the error type that we use internally for passing structured errors
+type VitessError struct {
+	// Error code of the Vitess error
+	Code int
+	// Additional context string, distinct from the error message. For example, if
+	// you wanted an error like "foo error: original error", the Message string
+	// should be "foo error: "
+	Message string
+	err     error
+}
+
+// Error implements the error interface. For now, it should exactly recreate the original error string.
+// It intentionally (for now) does not expose all the information that VitessError has. This
+// is so that it can be used in the mixed state where parts of the stack are trying to parse
+// error strings.
+func (e *VitessError) Error() string {
+	return fmt.Sprintf("%v", e.err)
+}
+
+// AsString returns a VitessError as a string, with more detailed information than Error().
```

Comment:
I explicitly don't want anything to default to using this String method. Naming it String() would cause fmt to convert it to this string, which I'd prefer not to do for now. The only default conversion to a string that I want is in a way that would make it consistent with the previous error strings.